usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -4601,6 +4612,245 @@ static CK_RV rsa_ec_generate_keypair(STDLL_TokData_t * tokdata,
+                                     CK_SESSION_HANDLE h)
+{
+    ep11_private_data_t *ep11_data = tokdata->private_data;
+    CK_RV rc;
+    CK_ATTRIBUTE *attr = NULL;
+    CK_BYTE privkey_blob[MAX_BLOBSIZE];
+    size_t privkey_blob_len = sizeof(privkey_blob);
+    unsigned char spki[MAX_BLOBSIZE];
+    size_t spki_len = sizeof(spki);
+    CK_BYTE publblob[MAX_BLOBSIZE];
+    size_t publblobsize = sizeof(publblob);
+    CK_BYTE privblob[MAX_BLOBSIZE];
+    size_t privblobsize = sizeof(privblob);
+    CK_ULONG i;
+    CK_ULONG bit_str_len;
+    CK_BYTE *key;
+    CK_BYTE *data, *oid, *parm;
+    CK_ULONG data_len, oid_len, parm_len;
+    CK_ULONG field_len;
+    CK_ATTRIBUTE_PTR new_pPublicKeyTemplate = NULL;
+    CK_ULONG new_ulPublicKeyAttributeCount = 0;
+    CK_ATTRIBUTE_PTR new_pPrivateKeyTemplate = NULL;
+    CK_ULONG new_ulPrivateKeyAttributeCount = 0;
+    CK_ULONG ktype = CKK_IBM_PQC_DILITHIUM;
+    unsigned char *ep11_pin_blob = NULL;
+    CK_ULONG ep11_pin_blob_len = 0;
+    ep11_session_t *ep11_session = (ep11_session_t *) sess->private_data;
+    CK_BYTE *seed, *t1;
+
+    UNUSED(h);
+
+    if (pMechanism->mechanism != CKM_IBM_DILITHIUM) {
+        TRACE_ERROR("Invalid mechanism provided for %s\n ", __func__);
+        return CKR_MECHANISM_INVALID;
+    }
+
+    rc = check_key_attributes(tokdata, ktype, CKO_PUBLIC_KEY,
+                              pPublicKeyTemplate, ulPublicKeyAttributeCount,
+                              &new_pPublicKeyTemplate,
+                              &new_ulPublicKeyAttributeCount);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s Dilithium check public key attributes failed with "
+                    "rc=0x%lx\n", __func__, rc);
+        return rc;
+    }
+
+    rc = check_key_attributes(tokdata, ktype, CKO_PRIVATE_KEY,
+                              pPrivateKeyTemplate, ulPrivateKeyAttributeCount,
+                              &new_pPrivateKeyTemplate,
+                              &new_ulPrivateKeyAttributeCount);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s Dilithium check private key attributes failed with "
+                    "rc=0x%lx\n", __func__, rc);
+        goto error;
+    }
+
+    rc = override_key_attributes(tokdata, ktype, CKO_PUBLIC_KEY,
+                                 new_pPublicKeyTemplate,
+                                 new_ulPublicKeyAttributeCount);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s Dilithium override public key attributes failed with "
+                    "rc=0x%lx\n", __func__, rc);
+        goto error;
+    }
+
+    rc = override_key_attributes(tokdata, ktype, CKO_PRIVATE_KEY,
+                                 new_pPrivateKeyTemplate,
+                                 new_ulPrivateKeyAttributeCount);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("%s Dilithium override private key attributes failed with "
+                    "rc=0x%lx\n", __func__, rc);
+        goto error;
+    }
+
+    /* debug */
```

Comment:
did you intend to put it inside DEBUG ifdef ?

Reply:
This trace is indeed intended to be displayed with trace level 4, but we could also decide to move it to trace level 5 and put a #ifdef DEBUG around it. The same piece of code shows up in rsa_ec_generate_keypair(). So whatever we decide, we should at least have the same logic for these two code pieces.

Reply:
okay, just the "debug" comment made me think you intended to change something here